presto-hive/src/main/java/com/facebook/presto/hive/HiveStagingFileCommitter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.HdfsEnvironment.HdfsContext;
+import com.facebook.presto.hive.PartitionUpdate.StagingFileInfo;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.SchemaTableName;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILESYSTEM_ERROR;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HiveStagingFileCommitter
+        implements StagingFileCommitter
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final Executor renameExecutor;
+
+    public HiveStagingFileCommitter(
+            HdfsEnvironment hdfsEnvironment,
+            Executor renameExecutor)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.renameExecutor = requireNonNull(renameExecutor, "renameExecutor is null");
```

Comment:
Make sure this executor has a threads limit. We don't want to suddenly spawn 30k threads.

Reply:
I agree we don't want to have too many threads. But it should be done at the executor construction time right (i.e. the callee of this `HiveStagingFileCommitter` constructor? )

Reply:
Yeah. I mean we have to make sure that the `renameExecutor` has the threads count limited. Most likely it does, as it was introduced long time ago, and has been already used for renames.

Reply:
I think it is bounded because it is defined here (https://github.com/prestodb/presto/blob/a683e76a8258a7169d3c8fdf1cd8d9a3eaacc1e1/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataFactory.java#L141), and default value for `maxConcurrentFileRenames` is 20.